Enable multiQuery optimization from within parent steps #1044

smcquillan opened the pull request:
This pull request seeks to address issue #1044 by enabling the multiQuery optimization for JanusGraphVertexStep which are children of the RepeatStep, MatchStep, BranchStep, and OptionalStep and more.

As I mentioned in the issue, the parent step drip feeds the starts one at a time into its child traversals. Re-initialising for each new start is one option, but that means a trip to the storage backend for each new start and defeats the point of the multiQuery optimisation. The simplest way to get around this would be for the JanusGraphVertexStep to request all the starts from its parent and use those when doing its initialisation, but there's no public method which allows us do do this.

To get around the fact that tinkerpop doesn't expose the starts I've created a new step named `JanusGraphMultiQueryStep` which, like the profiling steps, is just a simple passthrough. It caches the starts sent to the parent step and later allows a child `JanusGraphVertexStep` to request those starts and fully initialise itself with a single multiQuery to the storage backend. 

I've been running custom builds with these changes against our application which makes heavy use of repeat, union, choose, filer and optional steps and am seeing significant performance improvements.

Signed-off-by: Scott McQuillan <scott.mcquillan@uk.ibm.com>

-----

Thank you for contributing to JanusGraph!

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there an issue associated with this PR? Is it referenced in the commit message?
- [x] Does your PR body contain #xyz where xyz is the issue number you are trying to resolve?
- [x] Has your PR been rebased against the latest commit within the target branch (typically `master`)?
- [x] Is your initial contribution a single, squashed commit?

### For code changes:
- [x] Have you written and/or updated unit tests to verify your changes?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [ ] If applicable, have you updated the LICENSE.txt file, including the main LICENSE.txt file in the root of this repository?
- [ ] If applicable, have you updated the NOTICE.txt file, including the main NOTICE.txt file found in the root of this repository?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?
- [ ] If this PR is a documentation-only change, have you added a `[skip ci]`
  tag to the first line of your commit message to avoid spending CPU cycles in
  Travis CI when no code, tests, or build configuration are modified?

### Note:
Please ensure that once the PR is submitted, you check Travis CI for build issues and submit an update to your PR as soon as possible.

porunov commented on janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/optimize/JanusGraphMultiQueryStep.java:
```diff
@@ -0,0 +1,75 @@
+            final List<Traverser.Admin<Vertex>> newStarters = new ArrayList<>();
+            starts.forEachRemaining(s -> { newStarters.add(s); });
+            starts.add(newStarters.iterator());
+            cachedStarts.addAll(newStarters);
```
I just think that it is an unnecessary loop. It isn't critical but I think we can omit it like here:
```
final List<Traverser.Admin<Vertex>> newStarters = new ArrayList<>();
while (starts.hasNext()){
    Admin<Vertex> vertexAdmin = starts.next();
    newStarters.add(vertexAdmin);
    cachedStarts.add(vertexAdmin);
}
starts.add(newStarters.iterator());
cachedStarts.add(start);
```
Also, could you explain the line `starts.add(newStarters.iterator());`? I am not sure why do we need it. I initially sought that you want to expend the `starts` iterator with same elements. Then I checked that the iterator uses `TraverserSet` for its elements. So, when you actually call `starts.add(newStarters.iterator());` it actually executes the code like `existing.merge(traverser);` because all elements from `newStarters` already exists in `starts`. So we actually can replace the code to be like:
```
while (starts.hasNext()){
    Admin<Vertex> vertexAdmin = starts.next();
    cachedStarts.add(vertexAdmin);
    vertexAdmin.merge(vertexAdmin);
}
cachedStarts.add(start);
```
Am I missing something?

porunov replied:
Also, I didn't check but I assume that `merge` method just rewrites all changes into the existing object. So, in current situation it does nothing (need to check). If so, the code can be simplified like:
```
while (starts.hasNext()){
    cachedStarts.add(starts.next());
}
cachedStarts.add(start);
```

porunov replied:
Sorry, I missed that `starts` iterator won't be reusable after that

porunov replied:
I think then the next code can be used:
```
final List<Traverser.Admin<Vertex>> newStarters = new ArrayList<>();
starts.forEachRemaining(s -> {
    newStarters.add(s);
    cachedStarts.add(s);
});
starts.add(newStarters.iterator());
cachedStarts.add(start);
```

smcquillan replied:
I followed the patter used in [JanusGraphVertexStep. initialize()](https://github.com/JanusGraph/janusgraph/blob/master/janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/optimize/JanusGraphVertexStep.java#L94) and it confused me at first too but as I think you've just spotted the [starts.next()](https://github.com/apache/tinkerpop/blob/master/gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/process/traversal/step/util/ExpandableStepIterator.java#L53) is actually removing the start. So after we've looked at all the starts we need to call `starts.add(newStarters.iterator());` to put them all back again. 

Have just pushed up a change for the other comments, let me know if you think we need changes here too.

smcquillan replied:
Think our comments crossed each other - your last one looks good will change it to do that.

smcquillan commented on janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/optimize/JanusGraphMultiQueryStep.java:
```diff
@@ -0,0 +1,75 @@
+        Admin<Vertex> start = this.starts.next();
+        if (!cachedStarts.contains(start))
+        {
+            if (cachedStartsAccessed == true) {
```
fixed and codacy-bot bot picked me up on this one too.

porunov commented on janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/optimize/JanusGraphVertexStep.java:
```diff
@@ -85,20 +89,68 @@ private void initialize() {
+            multiQueryResults = results;
+        } else {
+            multiQueryResults.putAll(results);
+            results.clear();
```
Same here. Just would like to know why do we call `clear` here.

smcquillan replied:
Good Question, and given that the results goes out of scope shortly afterwards we don't need it, I've removed it.

smcquillan commented on janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/optimize/JanusGraphPropertiesStep.java:
```diff
@@ -97,12 +95,32 @@ private void initialize() {
+        else
+        {
+            multiQueryResults.putAll(results);
+            results.clear();
```
Good Question, and given that the results goes out of scope shortly afterwards we don't need it, I've removed it.

FlorianHockmann commented:
@smcquillan I didn't find the time to review this yet, but I think that I found a way how we can test these kind of optimizations easier. Since these optimizations are applied as a traversal strategy that changes the traversal, it should be possible to compare an unoptimized traversal after the strategies were applied with an optimized traversal as we expect it. So, something like this:

```java
GraphTraversal t = g.V().repeat(out().hasId(3)).times(3);

t.asAdmin().applyStrategies();

GraphTraversal optimizedTraversal = // construct the traversal as you expect it after optimization
assertEquals(optimizedTraversal , t);
```

It may be necessary to construct the optimized traversal with something like `t.asAdmin().addStep()` as we probably can't add JanusGraph specific steps otherwise.

That's also how strategies are [tested in TinkerPop](https://github.com/apache/tinkerpop/blob/667134d608ca29828457cd3eebe365ee1fd330ce/gremlin-core/src/test/java/org/apache/tinkerpop/gremlin/process/traversal/strategy/optimization/EarlyLimitStrategyTest.java#L69). After looking for where we could add such a test in JanusGraph I realized that we already have that kind of tests in [JanusGraphStepStrategyTest](https://github.com/JanusGraph/janusgraph/blob/master/janusgraph-test/src/test/java/org/janusgraph/blueprints/process/traversal/strategy/optimization/JanusGraphStepStrategyTest.java).

Maybe this can help you here (and maybe in the other PR) to better test this, especially to verify in which scenarios the optimization kicks in.

smcquillan commented:
@FlorianHockmann thanks for the tip, I think that'll be a great help in verifying that the new step is appearing as expected in the correct parts of the traversal. I'll look to add some additional tests using this approach. I've also had some success in extending this approach to the **repeat** and **match** steps so will include tests for them too.

smcquillan commented on janusgraph-test/src/test/java/org/janusgraph/blueprints/process/traversal/strategy/optimization/JanusGraphStepStrategyTest.java:
```diff
@@ -191,4 +261,52 @@ public void doTest(Traversal original, Traversal optimized, Collection<Traversal
         });
     }
+    
+    private static Stream<Arguments> generateMultiQueryTestParameters() {
```
This method is used but via an annotation but it seems the codacy-bot wasn't able to pick that it.

smcquillan commented on janusgraph-test/src/test/java/org/janusgraph/blueprints/process/traversal/strategy/optimization/JanusGraphStepStrategyTest.java:
```diff
@@ -191,4 +261,52 @@ public void doTest(Traversal original, Traversal optimized, Collection<Traversal
         });
     }
+    
+    private static Stream<Arguments> generateMultiQueryTestParameters() {
```
This method is used via an annotation but it appears that the codacy-bot can't detect that.

smcquillan commented:
@FlorianHockmann I’ve pushed up some additional changes in a new commit so people can see the differences to original pull request as reviewed by @porunov. In summary the additional changes are:

- JanusGraphMultiQueryStep now stores the class name of the parent traversal step it supports, which makes the profiling data clearer
- JanusGraphVertexStep now caches the JanusGraphMultiQueryStep (if found) so it doesn’t need to look it up again each time
- Side effect type steps are no longer counted when checking to see if the first step of a child traversal is a JanusGraphVertexStep
- RepeatStep is now supported, but unlike the other steps the JanusGraphMultiQueryStep goes at the end so it can feed its starts into the next iteration

In addition, as suggested, I’ve added a set of tests for this new optimisation strategy into JanusGraphStepStrategyTest. This was a worthwhile exercise as it led to my realisation that a simple `.as("name")` side effect at the start of a child traversal would have prevented the optimisation kicking in so fixed that and added a test for it. 

With regards to the support for MatchStep I’ve given up on that one. The reason being that while we could feed the starts into all its child traversals it’s not appropriate for all of them given that some children are actually fed their starts in from other children of the MatchStep. On the bright side however, this pull request does now allow the use of multiQuery within the child traversals of the match step, so it’s still an improvement over what we had before. 

I should also add that I’ve not been testing these changes in isolation. I’ve also been creating custom builds of the application I work on which include these changes and have seen the query times for some traversals drop by half.

smcquillan commented:
Amended the commit to make some minor changes to comments and white space and remove an unnecessary condition.

smcquillan commented:
@FlorianHockmann FYI I made a some minor changes today to fix some white space and comments, plus remove one redundant check. But the build seems to have failed, I assume for spurious reasons as I don't see how the changes I made could have caused that.

porunov commented:
@smcquillan It is my mistake caused by #1506 . I will merge #1523 now with CTR to fix the problem. You will have to just rebase against the latest commit

porunov commented:
@smcquillan I've merged the fix. Now you can rebase against the latest commit in master and the build should pass.

smcquillan commented on janusgraph-test/src/test/java/org/janusgraph/blueprints/process/traversal/strategy/optimization/JanusGraphStepStrategyTest.java:
```diff
@@ -191,4 +261,52 @@ public void doTest(Traversal original, Traversal optimized, Collection<Traversal
         });
     }
+    
+    private static Stream<Arguments> generateMultiQueryTestParameters() {
```
Method is used, but via an annotation and codacy-bot can't detect that.

smcquillan commented:
As I was expecting, got some conflicts now that https://github.com/JanusGraph/janusgraph/pull/1460 has been merged, will resolve them an push up a fix shortly.

smcquillan commented on janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/optimize/JanusGraphTraversalUtil.java:
```diff
@@ -85,4 +108,45 @@ public static JanusGraphTransaction getTx(Traversal.Admin<?, ?> traversal) {
+        }
+        return Lists.newArrayList(steps);
+    }
+    private static void getMultiQueryCompatibleStepsFromChildTraversal(Traversal.Admin<?,?> traversal, Step step, Set<Step> steps) {
```
Fair nitpick, have renamed these parameters to make things clearer

smcquillan commented:
Have rebased against develop and squashed things into a single commit
